onnxruntime/core/mlas/lib/amd64/QgemmU8S8KernelAmx.asm

```diff
@@ -0,0 +1,554 @@
+;++
+;
+;Copyright (c) Microsoft Corporation. All rights reserved.
+;
+;Licensed under the MIT License.
+;
+;Module Name:
+;
+;    QgemmU8S8KernelAmx.asm
+;
+;Abstract:
+;
+;    This module implements the packing functions for the quantized integer matrix/matrix
+;    multiply operation (QGEMM).
+;
+;    These packing functions are suited for AMX Qgemm kernel. The implementation only
+;    uses AVX2 instructions.
+;
+;--
+
+        .xlist
+INCLUDE mlasi.inc
+        .list
+
+
+;
+; Stack frame layout for the U8S8 CopyPackB routine.
+;
+
+GemmU8S8CopyPackBFrame STRUCT
+        SavedR12 QWORD ?
+        SavedRdi QWORD ?
+        SavedRsi QWORD ?
+        SavedRbx QWORD ?
+        SavedRbp QWORD ?
+        ReturnAddress QWORD ?
+        PreviousP1Home QWORD ?
+        PreviousP2Home QWORD ?
+        PreviousP3Home QWORD ?
+        PreviousP4Home QWORD ?
+        CountK QWORD ?
+        ColumnSumBuffer QWORD ?
+        BIsSigned QWORD ?
+
+GemmU8S8CopyPackBFrame ENDS
+
+;++
+;
+; Routine Description:
+;
+;   This routine copies elements from the source matrix to the destination
+;   packed buffer.
+;
+; Arguments:
+;
+;   D (rcx) - Supplies the address of the destination packed buffer.
+;
+;   B (rdx) - Supplies the address of the source matrix.
+;
+;   ldb (r8) - Supplies the number of elements per row of the source matrix.
+;
+;   CountN (r9) - Supplies the number of columns of the source matrix to copy.
+;
+;   CountK - Supplies the number of rows of the source matrix to copy.
+;
+;   ColumnSumBuffer - Supplies the address of the buffer to receive the sums of
+;       the elements along each of the columns.
+;
+;   BIsSigned - Supplies true if the source matrix is signed data, else false
+;       if the source matrix is unsigned data.
+;
+; Return Value:
+;
+;   None.
+;
+;--
+
+        NESTED_ENTRY MlasGemmU8S8CopyPackBAmx, _TEXT
+        rex_push_reg rbp
+        push_reg rbx
+        push_reg rsi
+        push_reg rdi
+        push_reg r12
+        END_PROLOGUE
+
+        mov     rsi,rdx                     ; Save B
+        lea     rdi,[r8+r8*2]               ; compute ldb * 3
+        mov     r10,GemmU8S8CopyPackBFrame.CountK[rsp]
+        mov     r11,GemmU8S8CopyPackBFrame.ColumnSumBuffer[rsp]
+        lea     r12,[r10+3]                 ; compute extra padding for 64|K
+        shr     r12,2
+        neg     r12
+        and     r12,15
+        vpcmpeqw ymm0,ymm0,ymm0             ; generate word vector [0xFFFF]
+        vpsrlw  ymm0,ymm0,15                ; generate word vector [0x0001]
+        vpsllw  ymm1,ymm0,8                 ; generate word vector [0x0100]
+        vpor    ymm1,ymm0,ymm1              ; generate word vector [0x0101]
+
+;
+; Compute the bit flip vector to adjust input from U8 to S8.
+;
+
+        vpxor   xmm2,xmm2,xmm2              ; generate word vector [0x0000]
```

Comment:
>vpxor   xmm2,xmm2,xmm2              ; generate word vector [0x0000] [](http://example.com/codeflow?start=8&length=67)

nit: there's no need to maintain a xmm2 with 0. instruction below 'vmovaps xmm17,xmm2' can use vpxor to clear the register.

Reply:
it's actually clearing out ymm2, which is either 0x0000 or 0x0808 based on whether B is signed or not. This is used for optional 8b top bit flip in line 144 and 145

Reply:
I see. Missed it.